gk/main.c

```diff
@@ -2148,52 +2156,83 @@ add_ggu_policy_bulk(struct gk_instance *instance, struct gk_config *gk_conf)
 		r = done_lookups + num_keys;
 		for (i = done_lookups; i < r; i++) {
 			int pos = pos_arr[i];
-			struct flow_entry *fe;
-			struct ggu_policy *policy = policies[i];
 
-			if (pos >= 0)
-				fe = &instance->ip_flow_entry_table[pos];
-			else if (instance->num_scan_del > 0)
+			if (pos >= 0) {
+				struct ggu_policy *policy = policies[i];
+				struct flow_entry *fe =
+					&instance->ip_flow_entry_table[pos];
+
+				update_flow_entry(fe, policy);
+			} else if (instance->num_scan_del > 0)
 				continue;
-			else {
-				struct gk_fib *fib = look_up_fib(
-					&gk_conf->lpm_tbl, &policy->flow);
-				if (fib == NULL || fib->action !=
-						GK_FWD_GRANTOR) {
-					/*
-					 * Drop this solicitation to add
-					 * a policy decision.
-					 */
-					char err_msg[128];
-					ret = snprintf(err_msg, sizeof(err_msg),
-						"gk: at %s initialize flow entry error",
-						__func__);
-					RTE_VERIFY(ret > 0 &&
-						ret < (int)sizeof(err_msg));
-					print_flow_err_msg(
-						&policy->flow, err_msg);
-					continue;
-				}
-
-				/*
-				 * The function add_ggu_policy_bulk() only
-				 * fills up GK_GRANTED and GK_DECLINED states.
-				 * So, it doesn't need to call
-				 * initialize_flow_entry().
-				 */
-				fe = add_new_flow_from_policy(fib,
-					policy, instance, gk_conf,
-					flow_hash_val_arr[i]);
-				if (fe == NULL)
-					continue;
+			else if (flow_arr[i]->proto == RTE_ETHER_TYPE_IPV4) {
+				lpm_lookup_pos[num_lpm_lookups] = i;
+				flows[num_lpm_lookups] = flow_arr[i];
+				num_lpm_lookups++;
+			} else {
+				lpm6_lookup_pos[num_lpm6_lookups] = i;
+				flows6[num_lpm6_lookups] = flow_arr[i];
+				num_lpm6_lookups++;
 			}
-
-			update_flow_entry(fe, policy);
 		}
 
 		done_lookups += num_keys;
 	}
 
+	/* The remaining flows need LPM lookups. */
+	lookup_fib_bulk(&gk_conf->lpm_tbl, flows, num_lpm_lookups, fibs);
+	lookup_fib6_bulk(&gk_conf->lpm_tbl, flows6, num_lpm6_lookups, fibs6);
```

Comment:
Due to the `else if (instance->num_scan_del > 0)` above these lines, it will be common for the number of lookups be zero. Add an early return to `lookup_fib_bulk()` and `lookup_fib6_bulk()` when there's no flow to look up.